mempool/src/pool/mod.rs

```diff
@@ -636,8 +636,9 @@ where
     pub async fn add_transaction(&mut self, tx: SignedTransaction) -> Result<(), Error> {
         let conflicts = self.validate_transaction(&tx).await?;
         self.store.drop_conflicts(conflicts);
-        self.finalize_tx(tx).await?;
+        self.finalize_tx(tx.clone()).await?;
         self.store.assert_valid();
+        self.events_controller.broadcast(MempoolEvent::NewTransaction(tx));
```

Comment:
This is wrong.

There seems to be a misunderstanding on why a NewTransaction event would have to exist, even though, speaking from a design point of you, it's the wrong thing to do. Transactions never emanate from the mempool. They end up there. They either come from peers or they come from rpc. When we submit a transaction through RPC, we are required to verify the validity of the transaction as if we're receiving it from a peer (to avoid being banned by peers) AND we are required to add it to the mempool (so that it can become part of a block) AND we're required to broadcast it to other peers (so that if we fail in adding it to a block, someone else will). Given that the first two requirements are done by the mempool, one would be tempted to think that the mempool should also do the broadcasting, but this is wrong. In fact, seeing this, I'm betting that the commig that has "deadlock" in the label was because of this.

So, two separate events have to happen in the RPC call:
1. Transaction should be added to the mempool (where it's also verified against the best chain)
2. Broadcast the transaction to the network.

It seems to me it's more reasonable to put both of these in the `submit_transaction` function of rpc, instead of handing transactions to the mempool and then get the mempool to broadcast them again.